Release Godot 4.2 stable and publish release notes (#758)
godotengine/godot-website: collections/_article/godot-4-2-arrives-in-style.md

---
title: "Godot 4.2 arrives in style!"

> excerpt: "More rendering features, better platform support, faster, easier to use, and more reliable!"

Reviewer: I think we usually do a complete sentence for the excerpt
```suggestion
excerpt: "Godot 4.2 brings more rendering features and better platform support, it's also faster, easier to use, and more reliable!"
```

categories: ["release"]
author: "Godot contributors"
image: /storage/blog/covers/godot-4-2-arrives-in-style.webp
date: 2023-11-30 12:00:00
---

> Earlier this year we stepped into the future of Godot with the [release of Godot 4.0](/article/godot-4-0-sets-sail), and 4 months later we followed it up with [Godot 4.1](/article/godot-4-1-is-here) — full of new features and improvements. Today we are proud to announce the release of Godot 4.2, our latest and greatest offering to game developers looking for a free and open engine that inspires joy and creativity!

Reviewer: ```suggestion
Earlier this year we stepped into the future of Godot with the [release of Godot 4.0](/article/godot-4-0-sets-sail), and 4 months later we followed it up with [Godot 4.1](/article/godot-4-1-is-here) which was full of new features and improvements. Today we are proud to announce the release of Godot 4.2, our latest and greatest offering to game developers looking for a free and open engine that inspires joy and creativity!
```
Small grammar fix. Em dashes add a long pause, but the sentence has to continue.

> Over the course of the last 5 months **359 contributors** submitted more than **1800 improvements** for this release, and we cannot thank each of them enough for their dedication and relentless spirit, their trust in the project, and their generosity. We also must express our love for everyone who reports issues during the testing period and helps us make sure that thousands of Godot users get the best version of the engine we can collectively create.

Reviewer 2: ```suggestion
Over the course of the last 5 months **359 contributors** submitted more than **1800 improvements** for this release, and we cannot thank each of them enough for their dedication and relentless spirit, their trust in the project, and their generosity. We also must express our love for everyone who reported issues during the testing period and helped us make sure that thousands of Godot users get the best version of the engine we can collectively create.
```
Kept same tense as "contributors submitted".

Author: This is meant to refer to people who do this in any release, not this particular release. So I think the use of the present tense was appropriate?

But maybe it needs to be continuous? "everyone reporting ... and helping ..."?

Author: ```suggestion
Over the course of the last 5 months **359 contributors** submitted more than **1800 improvements** for this release, and we cannot thank each of them enough for their dedication and relentless spirit, their trust in the project, and their generosity. We also must express our love for everyone reporting issues during the testing period and helping us make sure that thousands of Godot users get the best version of the engine we can collectively create.
```

> Among the changes in this release are both necessary bugfixes and exciting new features, turning Godot 4 into an even better and more polished tool for you to realize your game and app ideas. There are many more improvements to come, as our contributors, apparently, never sleep, but for now we draw the line here.

Reviewer 2: "draw the line" is usually used in a negative context. I can't immediately think of a more upbeat alternative...

Author: Yeah, I get that. On the flip side, this is taken straight from the beta 1 announcement (as is most of the middle part), so it definitely didn't cause any problems there :)

Author: Will change to another idiom with less bad connotations.

> Please, read on to learn about the most exciting additions in Godot 4.2, or click the links below to download the engine and jump straight into action!

Reviewer 2: ```suggestion
Please read on to learn about the most exciting additions in Godot 4.2, or click the links below to download the engine and jump straight into action!
```

{% include articles/download_card.html version="4.2" release="stable" article=page %}

## Giving back

As a community project, Godot is maintained by the efforts of volunteers. Thanks to [user and company donations](https://godotengine.org/donate/) we are able to hire and sponsor a number of core contributors to work full-time and dedicate their undivided attention to the needs of the project.

> Please, consider supporting the project financially with a [recurring monthly donation](https://fund.godotengine.org/), if you are able. Godot continues to grow with each passing month, and we are committed to bring you reliable and feature-rich releases as often as we can, to facilitate creative work of game developers of all levels. Recurring donations make sure the project is sustainable and its development can be planned for years ahead.

Reviewer 2: ```suggestion
Please consider supporting the project financially with a [recurring monthly donation](https://fund.godotengine.org/), if you are able. Godot continues to grow with each passing month, and we are committed to bringing you reliable and feature-rich releases as often as we can, to facilitate creative work of game developers of all levels. Recurring donations make sure the project is sustainable and its development can be planned for years ahead.
```

> Besides financial support, you can also give back by: writing detailed bug reports, contributing to the code base, writing documentation, writing guides (for the docs or on your own website), and supporting others on various [community platforms](https://docs.godotengine.org/en/latest/community/channels.html) by answering questions and providing helpful tips.

Reviewer 2: ```suggestion
Besides financial support, you can also give back by: writing detailed bug reports, contributing to the code base, writing documentation, writing guides (for the [Godot docs](https://docs.godotengine.org/) or on your own website), and supporting others on various [community platforms](https://docs.godotengine.org/en/latest/community/channels.html) by answering questions and providing helpful tips.
```

-----

## Highlights

In the sections below we will cover some of the most impactful changes that come to Godot 4.2 compared to 4.1. If you are interested in a complete list of changes, please refer to the [interactive changelog](https://godotengine.github.io/godot-interactive-changelog/#4.2), which contains links to relevant commits and PRs for this and every other release.

- [Critical and breaking changes](#critical-and-breaking-changes)
- [Core and systems](#core-and-systems)
- [2D and 3D features](#2d-and-3d-features)
- [Editor and usability](#editor-and-usability)
- [GDExtension](#gdextension)
- [GUI and theming](#gui-and-theming)
- [Import and asset pipeline](#import-and-asset-pipeline)
- [Input](#input)
- [Multiplayer and networking](#multiplayer-and-networking)
- [Porting and platforms](#porting-and-platforms)
- [Rendering, particles, and shaders](#rendering-particles-and-shaders)
- [Scripting](#scripting)
- [XR](#xr)

### Critical and breaking changes

> Godot [loosely follows](https://docs.godotengine.org/en/stable/about/release_policy.html#doc-release-policy) the semantic versioning system, but with a project of this scale it's impossible to move forward without implementing some changes that affect the compatibility with previous releases. We believe that for most users the migration process should be smooth and most critical steps are going to be resolved automatically.

Reviewer 2: ```suggestion
Godot [loosely follows](https://docs.godotengine.org/en/stable/about/release_policy.html#doc-release-policy) the semantic versioning system, but with a project of this scale it's impossible to move forward without implementing some changes that affect compatibility with previous releases. We believe that for most users the migration process should be smooth and most critical steps are going to be resolved automatically.
```

> To be properly diligent we describe below all the changes that require your attention when considering an upgrade:

Reviewer 3: Just a suggestion, but it could be worth it to link https://docs.godotengine.org/en/stable/tutorials/migrating/upgrading_to_godot_4.2.html prominently here.

- The new `AnimationMixer` node is added as an intermediate class for `AnimationPlayer` and `AnimationTree` ([GH-80813](https://github.com/godotengine/godot/pull/80813)).

>   - This brings both existing nodes to a parity in features and behavior. For the most part this change should not affect existing projects. However, the capture update mode for tracks is no longer available due to various issues in the original implementation. We will be working on providing this or similar functionality in a future version of Godot, but for now you have to work around this limitation using tweens.

Reviewer 2: ```suggestion
  - This brings both existing nodes to a parity in features and behavior. For the most part this change should not affect existing projects. However, the capture update mode for tracks is no longer available due to various issues in the original implementation. We will be working on providing this or similar functionality in a future version of Godot, but for now you can work around this limitation using tweens.
```
Sounds a bit more positive!

Author: Well, you literally have to, no way around it. But I don't mind the change.

Reviewer 4: Could go for "but for now you will need to work around", possibly more positive but also less non-absolute as it is a necessity

Author: ```suggestion
  - This brings both existing nodes to a parity in features and behavior. For the most part this change should not affect existing projects. However, the capture update mode for tracks is no longer available due to various issues in the original implementation. We will be working on providing this or similar functionality in a future version of Godot, but for now you need to work around this limitation using tweens.
```

- `GraphEdit` and `GraphNode` nodes are reworked with multiple renames, temporary removal of comment nodes, and introduction of `GraphElement` ([GH-79307](https://github.com/godotengine/godot/pull/79307), [GH-79308](https://github.com/godotengine/godot/pull/79308), [GH-79311](https://github.com/godotengine/godot/pull/79311), [GH-81582](https://github.com/godotengine/godot/pull/81582)).
  - For the time being, these nodes remain marked as experimental to gather further feedback from users.
- Several changes in the GLTF importer ([GH-80270](https://github.com/godotengine/godot/pull/80270), [GH-81264](https://github.com/godotengine/godot/pull/81264)) can lead to nodes in imported scenes having slightly different names, compared to Godot 4.1 and before.
  - Existing imported scenes will remain in the compatibility mode with their node names intact, until you decide to upgrade and reimport them.
- Changes to the way the engine stores meshes internally have been implemented to improve performance and bandwidth utilization on lower end devices ([GH-81138](https://github.com/godotengine/godot/pull/81138)).

>   - This is done with a grace system in place — existing meshes from Godot 4.1 and prior will be loaded and converted on the fly. There is also the new **Upgrade Mesh Surfaces** tool in the **Project > Tools** menu which will perform a one-time permanent conversion of all meshes in your project, which we recommend you to do.

Reviewer 2: ```suggestion
  - This is done with a grace system in place — existing meshes from Godot 4.1 and prior will be loaded and converted on the fly. There is also the new **Upgrade Mesh Surfaces** tool in the **Project > Tools** menu which will perform a one-time permanent conversion of all meshes in your project, which we recommend you do.
```

>   - This is done with a grace system in place — existing meshes from Godot 4.1 and prior will be loaded and converted on the fly. There is also the new **Upgrade Mesh Surfaces** tool in the **Project > Tools** menu which will perform a one-time permanent conversion of all meshes in your project, which we recommend you do.

Reviewer: ```suggestion
  - This is done with a grace system in place — existing meshes from Godot 4.1 and prior will be loaded and converted on the fly. There is also the new **Upgrade Mesh Surfaces** tool in the **Project > Tools** menu which will perform a one-time permanent conversion of all meshes in your project, which we recommend you do. Once upgraded to the new format, meshes will not be compatible with pre-4.2 releases.
```

Author: Edited your suggestion to fix a typo

- An optimization that required splitting raster barriers into vertex and fragment components changes relevant enumerations ([GH-77420](https://github.com/godotengine/godot/pull/77420)).
  - It's unlikely that this affects your project, unless you rely on specific numeric values and store them somewhere.

> - An internal rework of the `change_scene_to_*` methods introduces a change in the behavior compared to Godot 4.1 ([GH-78988](https://github.com/godotengine/godot/pull/78988)).

Reviewer 2: ```suggestion
- An internal rework of the `change_scene_to_*` methods introduces a change in behavior compared to Godot 4.1 ([GH-78988](https://github.com/godotengine/godot/pull/78988)).
```

>   - While previously the transition would happen at the end of the current frame, starting with Godot 4.2 the current scene is removed from the tree immediately. This means there is a brief period when no `current_scene` is present in the tree, so you may need to adjust your scripts to account for that. See the updated documentation for more details ([GH-85279](https://github.com/godotengine/godot/pull/85279)).

Reviewer 2: ```suggestion
  - While previously the transition would happen at the end of the current frame, starting with Godot 4.2 the current scene is removed from the tree immediately. This means there is a brief period when no `current_scene` is present in the tree, so you may need to adjust your scripts to account for that. See the [updated documentation](https://docs.godotengine.org/en/latest/classes/class_scenetree.html) for more details ([GH-85279](https://github.com/godotengine/godot/pull/85279)).
```
I'm not 100% sure this link is ideal, as `latest` will change in the future. If the new docs will be released at the same time as this PR, then pointing to `current` would be better I guess. In any case, I thought it would be nice to link to the docs, seeing as the user is being directed to them.

Reviewer 2: Mind you I've just noticed the link to `latest` docs in the Android section below so maybe that's OK here too.

Author: The latest link is fine, we don't have anything better until the version is released.

- Renaming an audio bus no longer emits the `bus_layout_changed` signal in `AudioServer`; you can use the new `bus_renamed` signal instead ([GH-81641](https://github.com/godotengine/godot/pull/81641)).
- The default base part of the Android package name was changed from `org.godotengine` to `com.example` ([GH-80761](https://github.com/godotengine/godot/pull/80761)).
- A previously unutilized notification, `NOTIFICATION_NODE_RECACHE_REQUESTED`, was completely removed from the engine ([GH-84419](https://github.com/godotengine/godot/pull/84419)).

### Core and systems

> Following the big changes to resource loading and multithreading done in Godot 4.1, the core team is more focused on stability this time around.

Reviewer 2: ```suggestion
Following the big changes to resource loading and multithreading done in Godot 4.1, the core team was more focused on stability this time around.
```

Author: I think either works, depending on what I want to say?

> - A previously unutilized notification, `NOTIFICATION_NODE_RECACHE_REQUESTED`, was completely removed from the engine ([GH-84419](https://github.com/godotengine/godot/pull/84419)).

Author: ```suggestion
- A previously unutilized notification, `NOTIFICATION_NODE_RECACHE_REQUESTED`, was completely removed from the engine ([GH-84419](https://github.com/godotengine/godot/pull/84419)).

To help you with the upgrade process Godot contributors are also working on a [migration guide for 4.2](https://docs.godotengine.org/en/4.2/tutorials/migrating/upgrading_to_godot_4.2.html). This is a live document and it will receive updates as they identify more areas deserving of your attention. You can help too by reporting missing steps to the [bug tracker](https://github.com/godotengine/godot-docs/issues).
```

These issues were fixed thanks to submissions from veteran contributors, such as [Pedro J. Estébanez](https://github.com/RandomShaper), [RedworkDE](https://github.com/RedworkDE), and [Rindbee](https://github.com/Rindbee), as well as project newcomers who only started contributing recently, [OXTyler](https://github.com/OXTyler) and [Jordyfel](https://github.com/Jordyfel).

A rather infamous issue, manifesting in crashes and exhaustive printing of `slot >= slot_max` to the output log, met its demise in Godot 4.2 at the hand of [Pāvels Nadtočajevs](https://github.com/bruvzg) ([GH-85280](https://github.com/godotengine/godot/pull/85280)). It was a pretty nasty bug which would only show up in export builds of your project and was related, one way or another, to manipulating the scene tree. And now it's thankfully gone.

> Sometimes a small mistake can result in major and unexpected differences in behavior. Such is the case with this bug that was spotted and fixed by the project's bugsquad powerhouse [AThousandShips](https://github.com/AThousandShips) ([GH-81037](https://github.com/godotengine/godot/pull/81037)). This issue caused abnormal memory usage and performance issues in seemingly trivial use cases. If you are using a lot of plain Godot objects, you should see a noticeable decrease in memory usage in 4.2.

Reviewer 4: Could we mention the future proofing one too as it's directly related to this? 🙂 being: https://github.com/godotengine/godot/pull/81020 (since it's talking about mistakes mentioning ensuring it won't happen again feels connected)

Author: Sure!

#### iOS

In turn, the iOS platform gets some love and parity with Android with the addition of one-click deploy ([GH-70662](https://github.com/godotengine/godot/pull/70662) by [Pāvels Nadtočajevs](https://github.com/bruvzg)). This feature means you can quickly deploy and remotely debug your project on an actual iOS device or an emulator with all the familiar editor tools at your disposal.

#### Linux

Linux support comes naturally in open source, as many Godot contributors daily-drive various Linux distributions. One of the more obscure variations that the engine supports is Linux for ARM processors. While the engine could always be built for this target, we never provided official distributions for Linux on ARM. But thanks to an effort from [Rémi Verschelde](https://github.com/akien-mga) to update our build toolchains ([GHBC-128](https://github.com/godotengine/build-containers/pull/128), [GHBC-131](https://github.com/godotengine/build-containers/pull/131)), we now can.

> You can find the link to the ARM version of the editor (both 32 and 64-bit) on the list of downloads on the [release page](/download/archive/4.2-stable). Export templates automatically include this platform. Since this is a relatively new version of the engine, some initial issues are highly likely. Please, treat this as an experimental feature and make sure to report problems if you run into any!

Reviewer 3: ```suggestion
Linux support comes naturally in open source, as many Godot contributors daily-drive various Linux distributions. One of the more obscure variations that the engine supports is Linux for ARM processors. While the engine could always be built for this target, we never provided official distributions for Linux on ARM. But thanks to an effort from [HP van Braam](https://github.com/hpvb) and [Rémi Verschelde](https://github.com/akien-mga) to update our build toolchains ([GHBC-128](https://github.com/godotengine/build-containers/pull/128), [GHBC-131](https://github.com/godotengine/build-containers/pull/131)), we now can.

You can find the link to the ARM version of the editor (both 32 and 64-bit) on the list of downloads on the [release page](/download/archive/4.2-stable). Export templates automatically include this platform. Since this is a relatively new version of the engine, some initial issues are highly likely – in fact, we're already aware of issues running executables on Raspberry Pi OS. Please, treat this as an experimental feature and make sure to report problems if you run into any!
```

Reviewer 3: Feel free to cut the second edit if it's too verbose - but HP did most of the work to get the new Linux toolchains :)

A prominent bug was fixed, which caused a crash if you went over 204 lights, decals, and reflection probes, combined ([GH-80845](https://github.com/godotengine/godot/pull/80845)). This issue stopped users from fully flexing the strength of the Forward+ renderer, but now they are free to use up to 512 of each entity type in the default configuration (this limit can be increased). And the built-in glow effect has been improved, making its visuals closer to the high quality look from Godot 3.x ([GH-82353](https://github.com/godotengine/godot/pull/82353) by [Raffaele Picca](https://github.com/RPicster)).

And now, for the big hitters...

#### New features

> Godot 4.2 features support for AMD's <abbr title="FidelityFX Super Resolution">FSR</abbr> 2.2, an open upscaling technique that works on GPUs of all vendors ([GH-81197](https://github.com/godotengine/godot/pull/81197)). This is possible thanks to [Darío Samo](https://github.com/DarioSamo), who also implemented a prerequisite support for motion vectors in skeletons, blend shapes, and particles ([GH-80618](https://github.com/godotengine/godot/pull/80618), [GH-80688](https://github.com/godotengine/godot/pull/80688)), and also solved many issues with the current <abbr title="Temporal anti-aliasing">TAA</abbr> system that had been plaguing users.

Reviewer 5: ```suggestion
Godot 4.2 features support for AMD's <abbr title="FidelityFX Super Resolution">FSR</abbr> 2.2, an open upscaling technique that works on GPUs of all vendors ([GH-81197](https://github.com/godotengine/godot/pull/81197)). This is possible thanks to [Darío Samo](https://github.com/DarioSamo), who also implemented a prerequisite support for motion vectors in skeletons, blend shapes, and particles ([GH-80618](https://github.com/godotengine/godot/pull/80618), [GH-80688](https://github.com/godotengine/godot/pull/80688)), and also solved many issues with the current <abbr title="Temporal anti-aliasing">TAA</abbr> system that had been plaguing users. You can try [this demo](https://github.com/todogodotcom/Garage42) to see <abbr title="FidelityFX Super Resolution">FSR</abbr> 2.2 in action.
```

Reviewer 5: This is how I would prompt people to try the demo to see the improvements it can make with their hardware.

<video autoplay loop muted playsinline title="Thousands of sparks affected by turbulence by passivestar">
  <source src="/storage/blog/godot-4-2-arrives-in-style/gpuparticles-by-passivestar.mp4" type="video/mp4">
</video>

Please note that the aforementioned changes to GPU particles mark the point where they start to deviate from CPU particles. This means that in the future more GPU-only features are going to be implemented, leaving CPU particles as a low-cost, simplified alternative, and not a direct 1-to-1 equivalent. GPU particles are supported by all rendering backends in Godot 4, and we encourage users to move towards using them. To aid in that transition, there is now a built-in tool to convert CPU particles to their GPU equivalent ([GH-80779](https://github.com/godotengine/godot/pull/80779) by [Yuri Rubinsky](https://github.com/Chaosus)).

#### Shaders

> Visual shaders got some nice improvements with the addition of drop-down list properties to custom nodes ([GH-81688](https://github.com/godotengine/godot/pull/81688) by [Yuri Rubinsky](https://github.com/Chaosus)), and output ports for vector types are now expandable by default ([GH-82088](https://github.com/godotengine/godot/pull/82088) by [Dennis Manaa](https://github.com/DennisManaa), their first contribution).

Reviewer 2: ```suggestion
Visual shaders got some nice improvements with the addition of drop-down list properties to custom nodes ([GH-81688](https://github.com/godotengine/godot/pull/81688) by [Yuri Rubinsky](https://github.com/Chaosus)), and output ports for vector types are now expandable by default ([GH-82088](https://github.com/godotengine/godot/pull/82088) by first-time contributor [Dennis Manaa](https://github.com/DennisManaa)).
```

Author: Why?

Besides typing improvements, another bunch of features from [Danil Alexeev](https://github.com/dalexeev) include raw string literals, or r-strings, ([GH-74995](https://github.com/godotengine/godot/pull/74995)), as well as return type covariance and parameter type contravariance ([GH-82477](https://github.com/godotengine/godot/pull/82477)). At the same time [George Marques](https://github.com/vnen) implemented pattern guards for more advanced `match` statements ([GH-80085](https://github.com/godotengine/godot/pull/80085)), and he also optimized operator calls
([GH-79990](https://github.com/godotengine/godot/pull/79990)).

> Regular and documentation comments are seeing some upgrades as well. Special TODO markers (such as `TODO`, `FIXME`, etc) are now highlighted in the code editor to better grab your attention ([GH-79761](https://github.com/godotengine/godot/pull/79761)). You can configure both keywords and colors that are given to them in the editor settings. Documentation comments now support `@deprecated` and `@experimental` tags to mark class members accordingly in the generated documentation ([GH-78941](https://github.com/godotengine/godot/pull/78941)). Documentation generation itself has also been improved ([GH-80745](https://github.com/godotengine/godot/pull/80745)). All these changes were submitted by, once again, [Danil Alexeev](https://github.com/dalexeev).

Reviewer 2: ```suggestion
Regular and documentation comments are seeing some upgrades as well. Special TODO markers (such as `TODO`, `FIXME`, etc.) are now highlighted in the code editor to better grab your attention ([GH-79761](https://github.com/godotengine/godot/pull/79761)). You can configure both keywords and colors that are given to them in the editor settings. Documentation comments now support `@deprecated` and `@experimental` tags to mark class members accordingly in the generated documentation ([GH-78941](https://github.com/godotengine/godot/pull/78941)). Documentation generation itself has also been improved ([GH-80745](https://github.com/godotengine/godot/pull/80745)). All these changes were submitted by, once again, [Danil Alexeev](https://github.com/dalexeev).
```

## Acknowledgements

Announcements such as this are as much of a team effort as the development of the engine itself. We look to our community for help with illustrating the changes and improvements that go into each release.

This article features screenshots and clips provided or based on the work of [GDQuest](https://www.gdquest.com/), [passivestar](https://twitter.com/passivestar_/status/1713487404349173958), [Minions Art](https://www.patreon.com/posts/stylized-skybox-89917325), [Icterus Games and Todogodot](https://github.com/todogodotcom/Garage42), as well as various engine contributors who provided illustrations with their pull requests.

The cover image was created using assets provided by [GDQuest](https://www.gdquest.com/) and [Thibaud](https://gotibo.fr/) ([Godot 4 3D Character pack](https://github.com/gdquest-demos/godot-4-3D-Characters/), [Godot 4 Stylized Sky shader](https://github.com/gdquest-demos/godot-4-stylized-sky)).

Make sure to support these creators and follow them for more amazing stuff!

## On to the next release!

> We already started work on Godot 4.3, which will come out at the start of the next year. We are dedicated to keeping up the pace and sticking to this reliable release cycle, so you never have to wait long for the next batch of improvements and new features.

Reviewer 2: ```suggestion
We already started work on Godot 4.3, which will come out at the start of next year. We are dedicated to keeping up the pace and sticking to this reliable release cycle, so you never have to wait long for the next batch of improvements and new features.
```

> We already started work on Godot 4.3, which will come out at the start of next year. We are dedicated to keeping up the pace and sticking to this reliable release cycle, so you never have to wait long for the next batch of improvements and new features.

Reviewer: ```suggestion
We already started work on Godot 4.3, which will come out early next year. We are dedicated to keeping up the pace and sticking to this reliable release cycle, so you never have to wait long for the next batch of improvements and new features.
```

"Start of next year" means January. We don't want to promise anything that early IMO. So being a bit more vague here will be safer
